### PR TITLE
glob dir entries with .ndignore

### PR DESCRIPTION
**Pull request by jrwren**

### CODE REVIEW

**Comment by deluan** [comment]
Hey, thanks for taking a stab at this, but I had a slightly different implementation in mind:
- `.ndignore` could be of two types: empty (meaning ignore this folder and all children) and containing globs (behave as .gitignore)
- `.ndignore` with globs could be found anywhere in the tree. In this case they would _add_ globs to the existing list (similar to what `git` does)

Seems that your solution takes care of the first item, but not the second one.... Do you think it is feasible to implement the second requirement as well?

> not Firewire

?? Didn't get it....

**Comment by jrwren** [comment]
For the second item, I'd like to do it, but I do not feel that a .ndignore deep in the tree should affect files in another part of the tree. So it isn't really a matter of adding globs to the existing list. It should only affect directory entries that are below that .ndignore. I didn't see an easy way of doing it without changing walking significantly and I didn't want to do that.

That said, maybe there is an easier way to do it and I don't see it. Maybe as easy as adding to the list when going down, but not coming back up, and maybe that is already there and just needs to be done?

Sorry, I'm rambling. I'll give it a try.

Sorry, 1394 is the IEEE standard for firewire. I was amused by issue 1394. I'm punchy today :).

**Comment by deluan** [comment]
Seems that this is working but it only matches folders, not files. I was hoping to use this to also ignore mixed versions of the same album. Imagine someone have a .FLAC and a .MP3 of the same album in the same folder, and they want ND to only scan FLAC files (this is a real request from a couple of users). If the ignore logic applied to files as well, they could have something like this in the root `.ndignore`: 
```
**/*.mp3
```

Unfortunately this is more complicated than just ignoring folders. I'm just pondering if it is worth to release this as is, and add the `ignore files` functionality later. I think it is ok, but we have to be sure that the syntax will be compatible...

Thoughts?

File: scanner/walk_dir_tree.go [hunk 1]
```diff
@@ -26,36 +28,39 @@ type (
...
 	}
 	close(results)
 	return err
 }
 
-func walkFolder(ctx context.Context, rootPath string, currentFolder string, results walkResults) error {
-	children, stats, err := loadDir(ctx, currentFolder)
+func walkFolder(ctx context.Context, rootPath string, currentFolder string, results walkResults, ignores []string) error {
+	// Look for .ndignore first
+	newignores := getIgnoreGlobs(currentFolder)
```
Comment (deluan) [inline scanner/walk_dir_tree.go:41]: Small nitpick: Please rename this var to proper camelCase (`newIgnores`(.

**Comment by jrwren** [comment]
Let me try some more. I admit I don't understand this code very well at all. I have just come to the understanding that this is the directory walker and deals with directories only and the rest is in tagscanner. I'll keep poking at it.

**Comment by jrwren** [comment]
Alright, I tried again. Thank you for your patience and for pointing out that it was not at all doing what I thought even after I wrote tests which in hindsight made no sense. I think that it is better now.

**Comment by joejoinerr** [comment]
Hey, any idea when this might be merged? Looking forward to this feature.

**Comment by pimzand** [comment]
I merged this PR into my fork, and the result does not work for me at all.
Nothing gets scanned into the db, even if no .ndignore file is present.
The process is consuming 106% CPU steadily.
strace shows continuous failures to open /.ndignore

**Comment by anpandey** [comment]
I think I fixed that with https://github.com/anpandey/navidrome/commit/a7bd344151f2d4f2fe45608fc2897c832297eef2 though I've yet to try it out on my actual music library.

**Comment by Whoisfoxmulder** [comment]
I have a folder with this file .ndignore. But in navidrome the albums are displayed anyway. Did a full rescan of the music library. What am I doing wrong?

image

**Comment by Deebster** [comment]
@Whoisfoxmulder your file is called `1.ndignore` instead of `.ndignore`

I remember that Windows used to have a problem with filenames that began with a dot; if this is affecting you, try naming the file with a dot at the end: `.ndignore.` - Windows will remove that last dot and allow the correct filename.

**Comment by Whoisfoxmulder** [comment]
Ah i see...yes now it works! Thanks a lot ❤️

**Pull request closed without merging**